### PR TITLE
Loop filter algorithmic optimizations

### PR DESCRIPTION
**Pull request by NaderMahdi**
**Algorithmic optimization of the loop-filter tools.**

**DLF-Filter**
 -   Loop filter Level 1 for encoder_modes 2 & 3
 -   Loop filter Level 2 for encoder_mode 1
 -   Loop filter Level 3 for encoder_mode 0

    // Loop filter Level                            Settings
    // 0                                            OFF
    // 1                                            CU-BASED
    // 2                                            LIGHT FRAME-BASED
    // 3                                            FULL FRAME-BASED

**CDEF-Filter**
 - cdef_filter_mode 1 for ENC_M3;
 - cdef_filter_mode 2 for ENC_M2;
 - cdef_filter_mode 3 for ENC_M1 & ENC_M2

	// CDEF Level                                   Settings
	// 0                                            OFF
	// 1                                            4 step refinement
	// 2                                            8 step refinement
	// 3                                            16 step refinement

**WIENER-Restoration**
 - wn_filter_mode 1 for ENC_M1,ENC_M2 and ENC_M3
 - wn_filter_mode 2 for ENC_M0

	// WN Level                               Settings
        // 1                                            5-Tap luma/ 5-Tap chroma
	// 2                                            7-Tap luma/ 5-Tap chroma

**SELF-GUIDED-restoration**
- sg_filter_mode = 1 for ENC_M3;
- sg_filter_mode = 2 for ENC_M2;
- sg_filter_mode = 3 for ENC_M1 & ENC_M0;

    // SG Level                                     Settings
    // 1                                                0 step refinement
    // 2                                                1 step refinement
    // 3                                                4 step refinement

**Bdrate/speed results**
svt-av1 | PSNR-Bdrate | SSIM-Bdrate | Speed Gain
-------- | -------------- | -------------- | ------------
M1       | 0.40%            | 0.20%            | 18.87%
M2       | 0.70%            | 0.40%            | 27.07%
M3       | 0.80%            | 0.60%            | 26.17%

### CODE REVIEW

File: Source/Lib/Codec/EbModeDecisionConfigurationProcess.c [hunk 1]
```diff
@@ -591,6 +591,74 @@ void SetSliceAndPictureChromaQpOffsets(
...
+	PictureControlSet_t                    *picture_control_set_ptr)
+{
+
+	Av1Common* cm = picture_control_set_ptr->parent_pcs_ptr->av1_cm;
+
+	memset(cm->sg_frame_ep_cnt, 0, SGRPROJ_PARAMS * sizeof(int32_t));
+	cm->sg_frame_ep = 0;
+
+	// NADER: set cm->sg_ref_frame_ep[0] = cm->sg_ref_frame_ep[1] = -1 to perform all iterations
+	if (picture_control_set_ptr->slice_type == I_SLICE) {
```
Comment (kylophone) [inline Source/Lib/Codec/EbModeDecisionConfigurationProcess.c]: Could be a switch. This is ok, too.

Comment (NaderMahdi) [reply Source/Lib/Codec/EbModeDecisionConfigurationProcess.c]: Done.

**Comment by OmranAbdelkafi** [comment]
Unexpected  significant BD-rate loss : Must be investigated first

File: Source/Lib/Codec/EbEncDecProcess.c [hunk 2]
```diff
@@ -1832,6 +1838,17 @@ void* EncDecKernel(void *input_ptr)
             }
 #endif
 #if !FILT_PROC
+#if FAST_SG
+            uint8_t best_ep_cnt = 0;
+            uint8_t best_ep = 0;
+            for (uint8_t i = 0; i < 16; i++) {
```
Comment (chkngit) [inline Source/Lib/Codec/EbEncDecProcess.c]: use predefined constant for the loop max bound 16, this should be the max number of SG param I guess!!

File: Source/Lib/Codec/EbModeDecisionConfigurationProcess.c [hunk 3]
```diff
@@ -591,6 +591,74 @@ void SetSliceAndPictureChromaQpOffsets(
...
+    PictureControlSet_t                    *picture_control_set_ptr)
+{
+
+    Av1Common* cm = picture_control_set_ptr->parent_pcs_ptr->av1_cm;
+    EbReferenceObject_t  * refObjL0, *refObjL1;
+    memset(cm->sg_frame_ep_cnt, 0, SGRPROJ_PARAMS * sizeof(int32_t));
+    cm->sg_frame_ep = 0;
+
+    // NADER: set cm->sg_ref_frame_ep[0] = cm->sg_ref_frame_ep[1] = -1 to perform all iterations
+    switch(picture_control_set_ptr->slice_type){
```
Comment (chkngit) [inline Source/Lib/Codec/EbModeDecisionConfigurationProcess.c]: slice type is HEVC specific, we should remove it as part of a whole encoder clean up task (not necessarily now)

File: Source/Lib/Codec/EbRestProcess.c [hunk 4]
```diff
@@ -331,6 +331,17 @@ void* rest_kernel(void *input_ptr)
                 cm->rst_info[2].frame_restoration_type = RESTORE_NONE;
             }
 
+#if FAST_SG
+            uint8_t best_ep_cnt = 0;
+            uint8_t best_ep = 0;
+            for (uint8_t i = 0; i < 16; i++) {
```
Comment (chkngit) [inline Source/Lib/Codec/EbRestProcess.c]: same comment for the number 16

File: Source/Lib/Codec/EbEncDecProcess.c (hunk 2 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbEncDecProcess.c]: Done

File: Source/Lib/Codec/EbRestProcess.c (hunk 4 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbRestProcess.c]: Done.

File: Source/Lib/Codec/EbModeDecisionConfigurationProcess.c (hunk 3 above)
Comment (NaderMahdi) [reply Source/Lib/Codec/EbModeDecisionConfigurationProcess.c]: I agree

**Comment by hassount** [comment]
@joelsole 
Any issues moving forward with this ?

**Comment by NaderMahdi** [comment]
@andrey-norkin @joelsole  

The merge of this PR is pending your approval. Could you please review it?

**Pull request merged**